File: backend/app/schemas/lifecycle.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.db.models import LifecycleAction, LifecycleHealthState


class LifecycleEvidence(BaseModel):
    market_view: str | None = None
    reason_codes: list[str] = []

class PositionLifecycleDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    portfolio_id: int
    instrument_id: int
    episode_number: int

    health_state: LifecycleHealthState
    recommended_action: LifecycleAction
    policy_version: str

    negative_confirmation_count: int
    strong_sell_confirmation_count: int
    recovery_confirmation_count: int
    add_confirmation_count: int

    last_counted_market_observation_key: str | None
    last_evaluated_at: datetime | None
    last_transition_at: datetime | None

    episode_started_at: datetime
    closed_at: datetime | None

    # Portfolio Context (derived dynamically when returned)
    symbol: str | None = None
    asset_class: str | None = None
    native_currency: str | None = None

    is_evaluable: bool | None = None
    data_state: str | None = None

    current_quantity: Decimal | None = None
    average_cost_base: Decimal | None = None
    current_market_value_base: Decimal | None = None
    unrealized_pnl_base: Decimal | None = None
    unrealized_pnl_pct: Decimal | None = None
    current_weight: Decimal | None = None

    market_score: Decimal | None = None
    technical_score: Decimal | None = None
    fundamental_score: Decimal | None = None
    data_quality_score: Decimal | None = None

    suggested_reduce_quantity: Decimal | None = None
    suggested_remaining_quantity: Decimal | None = None
    estimated_released_cash_base: Decimal | None = None

    evidence: LifecycleEvidence | None = None

class LifecycleSnapshotDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    episode_number: int
    health_state_before: LifecycleHealthState | None
    health_state_after: LifecycleHealthState
    recommended_action: LifecycleAction
    evaluated_at: datetime

    market_view: str | None
    reason_codes: str | None

class LifecycleSummaryDTO(BaseModel):
    open_positions_count: int
    stable_count: int
    watch_count: int
    reduce_count: int
    exit_count: int

File: frontend/types/lifecycle.ts
export enum LifecycleHealthState {
  STABLE = "STABLE",
  WATCH = "WATCH",
  CONFIRMED_DETERIORATION = "CONFIRMED_DETERIORATION",
  RECOVERING = "RECOVERING",
  CLOSED = "CLOSED",
}

export enum LifecycleAction {
  HOLD = "HOLD",
  CONSIDER_ADD = "CONSIDER_ADD",
  CONSIDER_REDUCE = "CONSIDER_REDUCE",
  CONSIDER_EXIT = "CONSIDER_EXIT",
  NO_ACTION_DATA = "NO_ACTION_DATA",
}

export interface LifecycleEvidence {
  market_view?: string;
  reason_codes: string[];
}

export interface PositionLifecycleDTO {
  id: number;
  portfolio_id: number;
  instrument_id: number;
  episode_number: number;
  
  health_state: LifecycleHealthState;
  recommended_action: LifecycleAction;
  policy_version: string;
  
  negative_confirmation_count: number;
  strong_sell_confirmation_count: number;
  recovery_confirmation_count: number;
  add_confirmation_count: number;
  
  last_counted_market_observation_key?: string;
  last_evaluated_at?: string;
  last_transition_at?: string;
  
  episode_started_at: string;
  closed_at?: string;
  
  symbol?: string;
  asset_class?: string;
  native_currency?: string;
  
  is_evaluable?: boolean;
  data_state?: string;
  
  current_quantity?: number | string;
  average_cost_base?: number | string;
  current_market_value_base?: number | string;
  unrealized_pnl_base?: number | string;
  unrealized_pnl_pct?: number | string;
  current_weight?: number | string;
  
  market_score?: number | string;
  technical_score?: number | string;
  fundamental_score?: number | string;
  data_quality_score?: number | string;
  
  suggested_reduce_quantity?: number | string;
  suggested_remaining_quantity?: number | string;
  estimated_released_cash_base?: number | string;
  
  evidence?: LifecycleEvidence;
}

export interface LifecycleSnapshotDTO {
  id: number;
  episode_number: number;
  health_state_before?: LifecycleHealthState;
  health_state_after: LifecycleHealthState;
  recommended_action: LifecycleAction;
  evaluated_at: string;
  
  market_view?: string;
  reason_codes?: string;
}

export interface LifecycleSummaryDTO {
  open_positions_count: number;
  stable_count: number;
  watch_count: number;
  reduce_count: number;
  exit_count: number;
}
